File: api_python/examples/500-Vision/01-vision.py
#! /usr/bin/env python3

###
# KINOVA (R) KORTEX (TM)
#
# Copyright (c) 2018 Kinova inc. All rights reserved.
#
# This software may be modified and distributed
# under the terms of the BSD 3-Clause license.
#
# Refer to the LICENSE file for details.
#
###

from kortex_api.UDPTransport import UDPTransport
from kortex_api.RouterClient import RouterClient, RouterClientSendOptions
from kortex_api.SessionManager import SessionManager

from kortex_api.autogen.client_stubs.VisionConfigClientRpc import VisionConfigClient
from kortex_api.autogen.client_stubs.DeviceConfigClientRpc import DeviceConfigClient
from kortex_api.autogen.client_stubs.DeviceManagerClientRpc import DeviceManagerClient

from kortex_api.autogen.messages import DeviceConfig_pb2, Session_pb2, DeviceManager_pb2, VisionConfig_pb2


def example_vision_config(device_manager_service, vision_config_service):

    # Get all device routing information (from DeviceManagerClient service)
    allDevicesInfo = device_manager_service.ReadAllDevices()

    vision_handles = [ hd for hd in allDevicesInfo.device_handle if hd.device_type == DeviceConfig_pb2.VISION ]
    if len(vision_handles) == 0:
        print("error: there is no vision device register in the devices infos")
    elif len(vision_handles) > 1:
        print("error: there are more than one vision device register in the devices infos")
    else:
        handle = vision_handles[0]

    sensor_id = VisionConfig_pb2.SensorIdentifier()
    sensor_id.sensor = VisionConfig_pb2.SENSOR_COLOR

    intrinsic_value = vision_config_service.GetIntrinsicParameters(sensor_id, handle.device_identifier)

    print("Width: {0}".format(intrinsic_value.width))
    print("Height: {0}".format(intrinsic_value.height))
    print("Principal point x: {0}".format(intrinsic_value.principal_point_x))
    print("Principal point y: {0}".format(intrinsic_value.principal_point_y))
    print("Focal length x: {0}".format(intrinsic_value.focal_length_x))
    print("Focal length y: {0}".format(intrinsic_value.focal_length_y))


if __name__ == "__main__":

    DEVICE_IP = "192.168.1.10"
    DEVICE_PORT = 10000

    # Setup API
    errorCallback = lambda kException: print("_________ callback error _________ {}".format(kException))
    transport = UDPTransport()
    router = RouterClient(transport, errorCallback)
    transport.connect(DEVICE_IP, DEVICE_PORT)

    # Create session
    session_info = Session_pb2.CreateSessionInfo()
    session_info.username = 'admin'
    session_info.password = 'admin'
    session_info.session_inactivity_timeout = 60000   # (milliseconds)
    session_info.connection_inactivity_timeout = 2000 # (milliseconds)

    session_manager = SessionManager(router)
    session_manager.CreateSession(session_info)

    # Create required services
    device_manager_service = DeviceManagerClient(router)
    vision_config_service = VisionConfigClient(router)

    # Example core
    example_vision_config(device_manager_service, vision_config_service)

    # Close API session
    session_manager.CloseSession()

    # Deactivate the router and cleanly disconnect from the transport object
    router.SetActivationStatus(False)
    transport.disconnect()

File: api_cpp/examples/500-Vision/01_vision.cpp
/*
* KINOVA (R) KORTEX (TM)
*
* Copyright (c) 2018 Kinova inc. All rights reserved.
*
* This software may be modified and distributed
* under the terms of the BSD 3-Clause license.
*
* Refer to the LICENSE file for details.
*
*/


#include <SessionManager.h>
#include <DeviceManagerClientRpc.h>
#include <VisionConfigClientRpc.h>

#include <RouterClient.h>
#include <TransportClientUdp.h>

namespace k_api = Kinova::Api;
namespace pb = google::protobuf;

#define PORT 10000
#define IP_ADDRESS "192.168.1.10"

void example_routed_vision(k_api::DeviceManager::DeviceManagerClient* pDeviceMng, k_api::VisionConfig::VisionConfigClient* pVision)
{
    // gets all device routing information (from DeviceManagerClient service)
    printf("-- ReadAllDevices --\n\n");
    auto allDevicesInfo = pDeviceMng->ReadAllDevices();

    // uses device routing information to route to every device (base, actuator, interconnect, etc.) in the armbase system and request general device information
    for ( auto dev : allDevicesInfo.device_handle() )
    {
        if ( dev.device_type() == k_api::Common::DeviceTypes::VISION )
        {
            printf("-- Using Vision Config Service to get intrinsic parameters --\n");
            k_api::VisionConfig::SensorIdentifier sensorId;
            sensorId.set_sensor(k_api::VisionConfig::SENSOR_COLOR);

            auto intrinsicValue = pVision->GetIntrinsicParameters(sensorId, dev.device_identifier());

            std::cout << "Width: " << intrinsicValue.width() << std::endl;
            std::cout << "Height: " << intrinsicValue.height() << std::endl;
            std::cout << "Principal point x: " << intrinsicValue.principal_point_x() << std::endl;
            std::cout << "Principal point y: " << intrinsicValue.principal_point_y() << std::endl;
            std::cout << "focal lenght x: " << intrinsicValue.focal_length_x() << std::endl;
            std::cout << "focal lenght y: " << intrinsicValue.focal_length_y() << std::endl;

            break;
        }
    }
}

int main(int argc, char **argv)
{
    auto pTransport = new k_api::TransportClientUdp();
    pTransport->connect(IP_ADDRESS, PORT);
    auto pRouterClient = new k_api::RouterClient(pTransport, [](k_api::KError err){ std::cout << "_________ callback error _________" << err.toString(); });

    // create session
    auto createSessionInfo = k_api::Session::CreateSessionInfo();
    createSessionInfo.set_username("admin");
    createSessionInfo.set_password("admin");
    createSessionInfo.set_session_inactivity_timeout(60000);
    createSessionInfo.set_connection_inactivity_timeout(2000);

    auto pSessionMng = new k_api::SessionManager(pRouterClient);
    pSessionMng->CreateSession(createSessionInfo);

    auto pDeviceMng = new k_api::DeviceManager::DeviceManagerClient(pRouterClient);
    auto pVision = new k_api::VisionConfig::VisionConfigClient(pRouterClient);

    example_routed_vision(pDeviceMng, pVision);

    pSessionMng->CloseSession();

    pTransport->disconnect();
    delete pSessionMng;
    delete pDeviceMng;
    delete pVision;
    delete pRouterClient;
    delete pTransport;
}